### ISSUE TITLE
MergeFlags() sometimes modifies its inputs in-place

### ISSUE THREAD

**Issue by ileonte** [SCons/scons#3665]
**Describe the bug**
`MergeFlags()` sometime modifies the data pointed to by its arguments in-place.

**Required information**
* Link to SCons Users thread discussing your issue: https://pairlist4.pair.net/pipermail/scons-users/2020-May/008152.html
* Version of SCons: **2.3.1, 2.5.1, 3.1.2**
* Version of Python: **2.7, 3.8**
* Which python distribution if applicable: **- (default install from distro packages)**
* How you installed SCons: **apt-get, portage, manual unpack of release tarballs**
* What Platform are you on?: **Linux - gentoo, Ubuntu 18.04, Ubuntu 20.04**
* How to reproduce your issue?  Please include a small self contained reproducer. Likely a SConstruct should do for most issues.

The absolute smallest example of the problem (that I can come up with) would be something like this:
```python    
flags = {
    'CFLAGS': ['-pipe', '-pthread', '-g'],
    'CXXFLAGS': ['${CFLAGS}', '-std=c++11'],
    'LINKFLAGS': ['-rdynamic', '-pthread'],
    'CPPDEFINES': ['DEF1=1', 'DEF2=2'],
}

print(flags)
env = Environment()
env.MergeFlags(flags)
print(flags)
```
This will print:
```
{'CFLAGS': ['-pipe', '-pthread', '-g'], 'CXXFLAGS': ['${CFLAGS}', '-std=c++11'], 'LINKFLAGS': ['-rdynamic', '-pthread'], 'CPPDEFINES': ['DEF1=1', 'DEF2=2']}
{'CFLAGS': ['-g', '-pthread', '-pipe'], 'CXXFLAGS': ['-std=c++11', '${CFLAGS}'], 'LINKFLAGS': ['-pthread', '-rdynamic'], 'CPPDEFINES': ['DEF2=2', 'DEF1=1']}
```
Notice how the lists are reversed (seemingly) in-place. This appears to happen for any key from the input dict that doesn't exist in the `Environment` at the time `MergeFlags()` is called. We can see this happening by modifying the example above to:
```python
flags = {
    'CFLAGS': ['-pipe', '-pthread', '-g'],
    'CXXFLAGS': ['${CFLAGS}', '-std=c++11'],
    'LINKFLAGS': ['-rdynamic', '-pthread'],
    'CPPDEFINES': ['DEF1=1', 'DEF2=2'],
}

print(flags)
env = Environment()
env['CFLAGS'] = ['-O1']   # set CFLAGS before MergeFlags() gets called
env.MergeFlags(flags)
print(flags)
```
This will now print:
```
{'CFLAGS': ['-pipe', '-pthread', '-g'], 'CXXFLAGS': ['${CFLAGS}', '-std=c++11'], 'LINKFLAGS': ['-rdynamic', '-pthread'], 'CPPDEFINES': ['DEF1=1', 'DEF2=2']}
{'CFLAGS': ['-pipe', '-pthread', '-g'], 'CXXFLAGS': ['-std=c++11', '${CFLAGS}'], 'LINKFLAGS': ['-pthread', '-rdynamic'], 'CPPDEFINES': ['DEF2=2', 'DEF1=1']}
```
You can now see that the list corresponding to the `CFLAGS` key doesn't change now that `env` contains the key when `MergeFlags()` is called. The other lists are still reversed in-place.

A slightly more involved example: https://raw.githubusercontent.com/ileonte/scons-bug-maybe/master/SConstruct
Running `scons` repeatedly with the above `SConstruct` will cause incorrect rebuilds of targets that are up-to-date. This happens because `MergeFlags()` modifies the `flags` dict in-place, thus causing the build commands to change because of parameter reordering.
* How you invoke scons (The command line you're using "scons --flags some_arguments"): `scons`

**Comment by mwichmann** [comment]
Can you try this quick patch?
```
diff --git a/SCons/Environment.py b/SCons/Environment.py
index 00379e185..db8af115d 100644
--- a/SCons/Environment.py
+++ b/SCons/Environment.py
@@ -855,8 +855,7 @@ class SubstitutionEnvironment(object):
                         t.append(v)
             else:
                 ### keep right-most occurence
-                orig.reverse()
-                for v in orig:
+                for v in orig[::-1]:
                     if v not in t:
                         t.insert(0, v)
             self[key] = t
```

**Comment by ileonte** [comment]
> Can you try this quick patch?
> 
> ```
> diff --git a/SCons/Environment.py b/SCons/Environment.py
> index 00379e185..db8af115d 100644
> --- a/SCons/Environment.py
> +++ b/SCons/Environment.py
> @@ -855,8 +855,7 @@ class SubstitutionEnvironment(object):
>                          t.append(v)
>              else:
>                  ### keep right-most occurence
> -                orig.reverse()
> -                for v in orig:
> +                for v in orig[::-1]:
>                      if v not in t:
>                          t.insert(0, v)
>              self[key] = t
> ```

I can confirm that this patch fixes the problem for me.

**Comment by bdbaddog** [comment]
Fixed by #3666 Merged into master
